### PR TITLE
ParticleMaterial: Sphere emission shape emitting from the volume.

### LINKED ISSUE #55398: Particle2D/3D emission shape "Sphere" is behaving different than documented.
### Godot version

4.0.dev 52b7d5f

### System information

Windows 10

### Issue description

The documentation about Particle2D / 3D emission shape mode "Sphere" is saying:
`Particles will be emitted in the volume of a sphere.`

The actual behaviour of the implementation (of 3.x and 4.0.dev) is different and particles are emitted from the surface instead of the volume.

While this behaviour could be useful and wanted to a user, usually an implementation using the volume of a sphere is more useful.

XeU2OutntS

It is especially visible and obvious in 2D because the z transformation is discarded, so there are more particles on the edges of the sphere.

### Steps to reproduce

- Create a Particle2D or 3D node
- Setup a simple particle system 
- Set it to Sphere Emission shape

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by RPicster**
Fixes #55398

It could be discussed if the current implementation of `sphere` could just be overwritten with the new behaviour.
In my opinion it would make sense to keep the current implementation in (imagine a magic orb emitting particles).

This should be backported if possible.

Old sphere emitter result:
XeU2OutntS

The new sphere emitter shape:
0N3lfPG6iA

### CODE REVIEW

**Comment by RPicster** [comment]
I added the same behaviour for CPUParticles2D and CPUParticles3D - So the behaviour and emission shapes are now consistent among all particle nodes.

**Comment by clayjohn** [comment]
The old behaviour isn't the surface of a sphere, it is uniform randomness within the volume of a sphere. Your implementation uses non-uniform randomness which focusses the points towards the center of the sphere. I.e. in your new code you get many more points close to the center of the sphere. 

I can't say which behaviour is preferable for the end user, but it isn't correct to say that the old behaviour generates particles on the surface of a sphere.

**Comment by RPicster** [comment]
I have to disagree. The old behaviour is the surface of a sphere. 
Definitely not uniform randomness within the volume of a sphere.
Please check a current build with CPUParticles3D, CPUParticles2D, GPUParticles3D and GPUParticles2D.

The only deviation is CPUParticles2D, but even there the distribution is not even and still increased on the edges.

And speaking as a very vfx-focused person, a uniform distribution inside a sphere is not what an artist wants because most of the time it is still emitted from the center.

Offering a uniform distribution would be another good option as a shape emitter, but in my experience is not needed very often.

**Comment by clayjohn** [comment]
You are correct. My bad! 

Indeed it makes sense to have volume emission as an option, and to match the behavior of box, the current behaviour should be renamed to sphere_surface. 

I will do a code review soon.

File: scene/3d/cpu_particles_3d.cpp [hunk 1]
```diff
@@ -792,6 +792,13 @@ void CPUParticles3D::_particles_process(double p_delta) {
 					//do none
 				} break;
 				case EMISSION_SHAPE_SPHERE: {
+					real_t s = 2.0 * Math::randf() - 1.0;
+					real_t t = Math_TAU * Math::randf();
+					real_t x = Math::randf();
+					real_t radius = emission_sphere_radius * Math::sqrt(1.0 - s * s);
+					p.transform.origin = Vector3(Math::lerp(0, radius * Math::cos(t), x), Math::lerp(0, radius * Math::sin(t), x), Math::lerp(0, emission_sphere_radius * s, x));
```
Comment (clayjohn) [inline scene/3d/cpu_particles_3d.cpp:799]: I think using ``Vector3.lerp`` is a little easier to read. IMO it makes it more obvious that we are interpolating between the sphere origin and the surface of the sphere by random value ``x``
```suggestion
					p.transform.origin = Vector3(0,0,0).lerp(Vector3(radius * Math::cos(t), radius * Math::sin(t), emission_sphere_radius * s), x);
```

Comment (RPicster) [reply scene/3d/cpu_particles_3d.cpp]: Thanks for your feedback, the suggested code is much easier to read!

**Comment by RPicster** [comment]
I changed the requested line. I would still love to get some feedback on this:

image

What would others think is the behaviour a user would expect?
Right or Middle or something in between?

**Comment by josephmbustamante** [comment]
> image

I think my preference is for the middle option. If I'm selecting a `sphere` emission shape, then I would expect it to simulate 3D volume, even in 2D, which means I think I would expect more particles where there would (ostensibly) be more volume. Besides the preference from a physics perspective though, I think I agree with Picster that middle option looks much more like an effect I want to create than the right option.

**Comment by GuyUnger** [comment]
i definitely think new and with easing make more sense that the old way, not sure which of those is better. i don't know what an even distribution of points in a sphere should look like or if it even matters if it's distributed perfectly equally.

about 2d, forgive me if i misunderstand the implementation, but if the 2d is just a flattened version of the 3d and you're planning to include both volume and rim, i would expect the rim in 2d to spawn on the rim of a circle and nowhere in the center, how is this planned to be implemented?

**Comment by djrain** [comment]
How about having just one "Sphere" shape, but adding a parameter that lets you adjust the distribution to your liking? 

https://user-images.githubusercontent.com/33777501/145101780-b940225d-6354-41ce-8ef1-b6290fd5f3d7.mov

When set to 0 you get a uniform distribution, -1 produces points only on the surface (current behavior), and positive values give increasingly more of a center bias.



Same thing can be applied to a 2D ring (haven't attempted 3D):


https://user-images.githubusercontent.com/33777501/145103173-c652fe77-6d00-45f4-8af2-da15c6224b89.mov

**Comment by RPicster** [comment]
> 

I really like that! We`ll talk on Discord and I will put it into the PR.

The bottom however would be the behaviour of the Ring Emitter that is already supplied.

**Review by clayjohn — APPROVED** [review]

**Comment by clayjohn** [comment]
@RPicster We discussed this at our rendering PR meeting. Consensus was to merge it as is. Having a slider for changing the distribution will add significant complexity and right now it seems like the greatest demand is either for emitting from volume, or emitting from surface.

**Comment by djrain** [comment]
> Having a slider for changing the distribution will add significant complexity and right now it seems like the greatest demand is either for emitting from volume, or emitting from surface.

I can understand the complexity issue, but if the team is open to adding some sort of distribution parameter in the near future... won't this change (having now two sphere shapes) only make it far more difficult to implement such a parameter later without breaking compatibility?

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**